server/athenian/api/experiments/preloading/entries.py

```diff
@@ -1,8 +1,137 @@
-from athenian.api.controllers.features.entries import \
-    MetricEntriesCalculator as OriginalMetricEntriesCalculator
+from datetime import datetime
+from typing import Collection, Optional, Set, Tuple
+
+import aiomcache
+import databases
+import pandas as pd
+
+from athenian.api.controllers.features.entries import (
+    MetricEntriesCalculator as OriginalMetricEntriesCalculator,
+)
+from athenian.api.controllers.miners.filters import JIRAFilter, LabelFilter
+from athenian.api.controllers.miners.github.pull_request import (
+    PullRequestMiner as OriginalPullRequestMiner,
+)
+from athenian.api.controllers.miners.types import PRParticipants, PRParticipationKind
+from athenian.api.models.metadata.github import PullRequest
+from athenian.api.tracing import sentry_span
+
+
+class PullRequestMiner(OriginalPullRequestMiner):
+    """Load all the information related to Pull Requests from the metadata DB. Iterate over it \
+    to access individual PR objects."""
+
+    @classmethod
+    @sentry_span
+    async def _fetch_prs_by_filters(
+        cls,
+        time_from: datetime,
+        time_to: datetime,
+        repositories: Set[str],
+        participants: PRParticipants,
+        labels: LabelFilter,
+        jira: JIRAFilter,
+        exclude_inactive: bool,
+        pr_blacklist: Optional[Collection[str]],
+        pr_whitelist: Optional[Collection[str]],
+        meta_ids: Tuple[int, ...],
+        mdb: databases.Database,
+        cache: Optional[aiomcache.Client],
+        columns=PullRequest,
+        updated_min: Optional[datetime] = None,
+        updated_max: Optional[datetime] = None,
+    ) -> pd.DataFrame:
+        if jira or labels:
+            # TODO: not supported yet, call original implementation
+            return await super(PullRequestMiner, cls)._fetch_prs_by_filters(
+                time_from,
+                time_to,
+                repositories,
+                participants,
+                labels,
+                jira,
+                exclude_inactive,
+                pr_blacklist,
+                pr_whitelist,
+                meta_ids,
+                mdb,
+                cache,
+                columns=columns,
+                updated_min=updated_min,
+                updated_max=updated_max,
+            )
+
+        assert (updated_min is None) == (updated_max is None)
+
+        df = mdb.cache.dfs["prs"].df
+
+        mask = (
+            (~df["closed"] | (df["closed_at"] >= time_from))
+            & (df["created_at"] < time_to)
+            & (df["acc_id"].isin(meta_ids))
+            & (~df["hidden"])
+            & (df["repository_full_name"].isin(repositories))
+        )
+
+        if pr_blacklist is not None:
+            pr_blacklist = [v.encode() for v in pr_blacklist]
+            mask &= ~df["node_id"].isin(pr_blacklist)
```

Comment:
Here, this will be way faster:
```suggestion
            mask &= np.in1d(df["node_id"].values.astype("S"), pr_blacklist, inverse=True)
```
provided `pr_blacklist` is already an "S" array.

Reply:
I'm wondering whether it makes sense to have these utils somewhere as we'll have a lot of `np.ind1d ...` calls.  Is this going to be faster even if `pr_blacklist` has to be converted to `"S"`? How much faster is it? Actually I could just measure it.

Reply:
"U" -> "S" conversion comes almost for free. What matters here is that both arrays become strongly typed vs. `object` and `np.in1d` takes the fastest path.